### DISCUSSION TITLE
Documentation examples

### DISCUSSION THREAD

**Discussion by oddlodd** [uktrade/stream-unzip#48] [General]
I think the project would benefit a lot from some actual examples.
Eg.:
Using request.get

Or using blobclient read to stream

Maybe i misunderstood this, but i need to download a zip from url or azure blob using python, uploading to a the blob,  but i dont understand how. Can i utilze the chunkcs from requests.get stream or blobclient's own stream?

**My specific example:**

 ```
**def download_file():
     url = 'someUrl'
     local_filename = f"someFilename.zip"
     with requests.get(url, stream=True, verify=False) as r:
         r.raise_for_status()
         blobclient= zipblob_service_client.get_blob_client(container, blob)
         for chunk in r.iter_content(chunk_size=1024 * 1024):
             blobclient.upload_blob(chunk, blob_type="AppendBlob")**
```

How would i go about updating it with stream-unzip to have the file unzipped before its written to blob

**Comment by michalc** [comment]
Hello 👋 

> I think the project would benefit a lot from some actual examples.

Agreed!

> Can i utilze the chunkcs from requests.get stream

Yes. So the response from `iter_content` is an iterable of `bytes` and so be passed to `stream_unzip`.

```python
with requests.get(url, stream=True) as r:
   r.raise_for_status()
   for name, size, chunks in stream_unzip(r.iter_content()):
       print(name)
       for chunk in chunks:
           print(chunk)
```

> or blobclient's own stream?

I think might have leave this to someone else to answer - I don't know much about Azure have to admit.

> How would i go about updating it with stream-unzip to have the file unzipped before its written to blob

But I'll take a guess at this one. Looks like upload_blob can take an iterable as its first argument https://learn.microsoft.com/en-us/python/api/azure-storage-blob/azure.storage.blob.blobclient?view=azure-python#azure-storage-blob-blobclient-upload-blob, and I guess it'll be an iterable of `bytes`. So you can pass it each `chunks`. I'm guessing a lot at what things mean, and this is untested, but the below might lead you somewhere? (I'm hoping it uploads all the members of a zip file...)

```python
with requests.get(url, stream=True) as r:
   r.raise_for_status()
   for name, size, chunks in stream_unzip(r.iter_content()):
        print(name)
        blobclient = zipblob_service_client.get_blob_client(container, name)
        blobclient.upload_blob(chunks, overwrite=True)
```
If you do get something working - you're very welcome to raise a PR with it adding it to the documentation. stream-zip, the sister library to this one, has "recipes" https://stream-zip.docs.trade.gov.uk/recipes/ . We could have a similar concept here?

**Comment by michalc** [comment]
> > or blobclient's own stream?

> I think might have leave this to someone else to answer - I don't know much about Azure have to admit.

Actually, I'm going to give this a go too. Looks like the return value of the "chunks" function is an iterable of bytes, and so can be passed to `stream-unzip` https://learn.microsoft.com/en-us/python/api/azure-storage-blob/azure.storage.blob.storagestreamdownloader?view=azure-python#azure-storage-blob-storagestreamdownloader-chunks

So it's still an untested guess, but to download a zip from one blob, and upload each member to its own blob, something like:

```python
stream = source_blob_client.download_blob()
for name, size, chunks in stream_unzip(stream.chunks()):
    print(name)
    blobclient = zipblob_service_client.get_blob_client(container, name)
    blobclient.upload_blob(chunks, overwrite=True)
```
To re-iterate - no idea if I'm using the Azure functions correctly.
